delta/plugins/archive/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/archive/ArchiveDownload.scala

```diff
@@ -0,0 +1,189 @@
+import ch.epfl.bluebrain.nexus.delta.plugins.archive.model.{ArchiveRejection, ArchiveResourceRepresentation, ArchiveValue}
+import ch.epfl.bluebrain.nexus.delta.plugins.storage.files.Files
+import ch.epfl.bluebrain.nexus.delta.plugins.storage.files.model.FileRejection
+import ch.epfl.bluebrain.nexus.delta.rdf.IriOrBNode.Iri
+import ch.epfl.bluebrain.nexus.delta.rdf.RdfError
+import ch.epfl.bluebrain.nexus.delta.rdf.implicits._
+import ch.epfl.bluebrain.nexus.delta.rdf.utils.JsonKeyOrdering
+import ch.epfl.bluebrain.nexus.delta.sdk.Permissions.resources
+import ch.epfl.bluebrain.nexus.delta.sdk.ReferenceExchange.ReferenceExchangeValue
+import ch.epfl.bluebrain.nexus.delta.sdk.model.IdSegment.IriSegment
+import ch.epfl.bluebrain.nexus.delta.sdk.model.ResourceRef
+import ch.epfl.bluebrain.nexus.delta.sdk.model.acls.AclAddressFilter.AnyOrganizationAnyProject
+import ch.epfl.bluebrain.nexus.delta.sdk.model.acls.{AclAddress, AclCollection}
+import ch.epfl.bluebrain.nexus.delta.sdk.model.identities.Caller
+import ch.epfl.bluebrain.nexus.delta.sdk.model.permissions.Permission
+import ch.epfl.bluebrain.nexus.delta.sdk.model.projects.ProjectRef
+import ch.epfl.bluebrain.nexus.delta.sdk.{Acls, AkkaSource, ReferenceExchange}
+import fs2.Stream
+import monix.bio.{IO, Task, UIO}
+
+import java.net.URLEncoder
+import java.nio.charset.StandardCharsets
+
+/**
+  * Archive download functionality.
+  */
+trait ArchiveDownload {
+
+  /**
+    * Generates an akka [[Source]] of bytes representing the tar.
+    *
+    * @param value          the archive value
+    * @param project        the archive parent project
+    * @param ignoreNotFound do not fail when resource references are not found
+    * @param caller         the caller to be used for checking for access
+    */
+  def apply(
+      value: ArchiveValue,
+      project: ProjectRef,
+      ignoreNotFound: Boolean
+  )(implicit caller: Caller): IO[ArchiveRejection, AkkaSource]
+
+}
+
+object ArchiveDownload {
+
+  /**
+    * The default [[ArchiveDownload]] implementation.
+    *
+    * @param exchanges the collection of [[ReferenceExchange]] implementations
+    * @param acls      the acls module
+    * @param files     the files module
+    * @param sort      the configuration for sorting json keys
+    */
+  class ArchiveDownloadImpl(exchanges: Set[ReferenceExchange], acls: Acls, files: Files)(implicit sort: JsonKeyOrdering)
+      extends ArchiveDownload {
+
+    override def apply(
+        value: ArchiveValue,
+        project: ProjectRef,
+        ignoreNotFound: Boolean
+    )(implicit caller: Caller): IO[ArchiveRejection, AkkaSource] = {
+      // the option allows recovery when trying to ignore resources that are not found
+      val listIO: IO[ArchiveRejection, List[Option[(TarArchiveMetadata, AkkaSource)]]] = {
+        // fetch all acls for the caller once for verifying for access
+        acls.listSelf(AnyOrganizationAnyProject(withAncestors = true)).flatMap { implicit aclCollection =>
```

Comment:
the acls should be checked using the existing `acls.authorizeForAny` where you pass all the projects in the `ResourceReference`.

Reply:
a ResourceReference defines a single project; instead of reading the acls of the caller every single time, the collection of acls is read only once for all projects and then checked against for every reference.

Reply:
the implementation of `authorizeForAny` is already optional, just checking once for each project

Reply:
the references are evaluated one at a time due to how files may be protected by custom permissions; every call to the acls.authorizeForAny would read the state of acls which is inefficient.
Alternately, if I were to partition the Set[ArchiveReference] to a Set[ResourceReference] and a Set[FileReference] the code would be much more complicated. I don't see any value in doing that.

Reply:
Yes, that's what I'm asking you to do. To just do a value.references.collect{...} (to collect the `Set[ResourceReference]` and then do a `acls. authorizeForAny` just once. That won't be inefficient.

Reply:
authorizeForAny does not make use of the cache and will recurse on fetchWithAncestors for each project entry. It's quite inefficient.

Reply:
This is the desired behaviour. Back then we said it is better to rely on the state since we are making sure there is no caching delay. 
It is not going to be inefficient since for ACL's we should always have the state in memory.